src/dataflow/src/adapter/error.rs

```diff
@@ -0,0 +1,112 @@
+use common_telemetry::common_error::status_code::StatusCode;
+use datatypes::data_type::ConcreteDataType;
+use servers::define_into_tonic_status;
+use snafu::{Location, Snafu};
+
+#[derive(Snafu)]
+#[snafu(visibility(pub))]
+#[stack_trace_debug]
+pub enum Error {
+    /// TODO(discord9): add detailed location of column
+    #[snafu(display("Stream Eval Failure, {}", raw))]
+    Eval { raw: EvalError },
+    #[snafu(display("Couldn't found table: {}", name))]
+    TableNotFound { name: String },
+    #[snafu(display("Table already exist: {}", name))]
+    TableAlreadyExist { name: String },
+    #[snafu(display("Failed to join task"))]
+    JoinTask {
+        #[snafu(source)]
+        error: tokio::task::JoinError,
+        location: Location,
+    },
+    #[snafu(display("Invalid query: {}", reason))]
+    InvalidQuery { reason: String, location: Location },
+}
+
+pub type Result<T> = std::result::Result<T, Error>;
+
+impl ErrorExt for Error {
+    fn status_code(&self) -> StatusCode {
+        match self {
+            Self::Eval { .. }
+            | &Self::TableNotFound { .. }
+            | &Self::TableAlreadyExist { .. }
+            | &Self::JoinTask { .. }
+            | &Self::InvalidQuery { .. } => StatusCode::Internal,
+        }
+    }
+
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+}
+
+define_into_tonic_status!(Error);
+
+use serde::{Deserialize, Serialize};
+
+// TODO(discord9): more error types
+#[derive(Ord, PartialOrd, Clone, Debug, Eq, Deserialize, Serialize, PartialEq, Hash)]
+pub enum DataflowError {
+    EvalError(Box<EvalError>),
+}
+
+impl From<EvalError> for DataflowError {
+    fn from(e: EvalError) -> Self {
+        DataflowError::EvalError(Box::new(e))
+    }
+}
+
+/// EvalError is about errors happen on columnar evaluation
+#[derive(Snafu)]
+#[snafu(visibility(pub))]
+//#[stack_trace_debug]
+#[derive(Ord, PartialOrd, Clone, Debug, Eq, Deserialize, Serialize, PartialEq, Hash)]
+pub enum EvalError {
+    #[snafu(display("Division by zero"))]
+    DivisionByZero,
+    #[snafu(display("Type mismatch: expected {}, actual {}", expected, actual))]
+    TypeMismatch {
+        expected: ConcreteDataType,
+        actual: ConcreteDataType,
+    },
+    /// can't nest datatypes error because EvalError need to be store in map and serialization
+    #[snafu(display("Fail to unpack from value to given type: {}", msg))]
+    TryFromValue { msg: String },
+    #[snafu(display(
+        "Fail to cast value of type {} to given type: {}. Detail: {}",
+        from,
+        to,
+        msg
+    ))]
+    CastValue {
+        from: ConcreteDataType,
+        to: ConcreteDataType,
+        msg: String,
+    },
+    #[snafu(display("Invalid argument: {}", reason))]
+    InvalidArgument { reason: String },
+    #[snafu(display("Internal error: {}", reason))]
+    Internal { reason: String },
+    #[snafu(display("Optimize error: {}", reason))]
+    Optimize { reason: String },
+}
```

Comment:
Is it possible to move to the expr mod? But in our practice, there is usually one Error for one crate. I don’t know if this is acceptable.

In addition, attribute seems to need to be improved
```rust
//#[stack_trace_debug]
#[derive(Ord, PartialOrd, Clone, Debug, Eq, Deserialize, Serialize, PartialEq, Hash)]
```

Reply:
> Is it possible to move to the expr mod? But in our practice, there is usually one Error for one crate. I don’t know if this is acceptable.
> 
> In addition, attribute seems to need to be improved
> 
> ```rust
> //#[stack_trace_debug]
> #[derive(Ord, PartialOrd, Clone, Debug, Eq, Deserialize, Serialize, PartialEq, Hash)]
> ```

`EvalError` is suppose to run through dataflow operator, and accumlate if multiple eval errors happens, and ser/deser to other nodes if necessary, so EvalError can't nest most other snafu error, but Error type require this flexibility, maybe just make `EvalError` pub(crate) and only expose `dataflow::Error`?